File: backend/service/time_utils.py
"""
Time Utility Functions for IST Timezone
Handles datetime formatting and duration calculations
"""
from datetime import datetime
from pytz import timezone
from typing import Optional

IST = timezone('Asia/Kolkata')


def format_ist(dt: datetime) -> str:
    """
    Format datetime to IST string.
    
    Args:
        dt: datetime object (can be naive or aware)
    
    Returns:
        Formatted string: "14-Jan-2026 03:30:45 PM IST"
    """
    if dt is None:
        return ""
    
    if dt.tzinfo is None:
        # Naive datetime, assume UTC
        dt = timezone('UTC').localize(dt)
    
    # Convert to IST
    ist_time = dt.astimezone(IST)
    
    # Format: 14-Jan-2026 03:30:45 PM IST
    return ist_time.strftime('%d-%b-%Y %I:%M:%S %p IST')


def format_duration(start_dt: datetime, end_dt: datetime) -> str:
    """
    Calculate and format duration between two datetimes.
    
    Args:
        start_dt: Start datetime
        end_dt: End datetime
    
    Returns:
        Formatted duration: "2h 15m 30s" or "10m 30s" or "45s"
    """
    if start_dt is None or end_dt is None:
        return "0s"
    
    diff = end_dt - start_dt
    total_seconds = int(diff.total_seconds())
    
    if total_seconds < 0:
        return "0s"
    
    hours = total_seconds // 3600
    minutes = (total_seconds % 3600) // 60
    seconds = total_seconds % 60
    
    if hours > 0:
        return f"{hours}h {minutes}m {seconds}s"
    elif minutes > 0:
        return f"{minutes}m {seconds}s"
    else:
        return f"{seconds}s"


def now_ist() -> datetime:
    """
    Get current datetime in IST.
    
    Returns:
        Current datetime in Asia/Kolkata timezone
    """
    return datetime.now(IST)


def current_ist_str() -> str:
    """
    Get current datetime formatted as IST string.
    
    Returns:
        Current time formatted as IST string
    """
    return format_ist(now_ist())

File: frontend/lib/timeUtils.ts
/**
 * Time Utility Functions for IST Timezone
 * Handles timestamp formatting and duration calculations
 */

const TIMEZONE = 'Asia/Kolkata';

/**
 * Format timestamp to IST string
 * @param timestamp - Unix timestamp in milliseconds or Date object
 * @returns Formatted string: "14-Jan-2026 03:30:45 PM IST"
 */
export function formatIST(timestamp: number | Date): string {
  const date = timestamp instanceof Date ? timestamp : new Date(timestamp);
  
  // Format: dd-MMM-yyyy hh:mm:ss AM/PM IST
  const options: Intl.DateTimeFormatOptions = {
    timeZone: TIMEZONE,
    day: '2-digit',
    month: 'short',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: true
  };
  
  const formatted = new Intl.DateTimeFormat('en-IN', options).format(date);
  
  // Convert to desired format: "14-Jan-2026 03:30:45 PM IST"
  // The Intl.DateTimeFormat gives us something like: "14 Jan 2026, 03:30:45 pm"
  // We need to clean it up
  const parts = formatted.split(', ');
  const datePart = parts[0].replace(/ /g, '-');
  const timePart = parts[1].toUpperCase();
  
  return `${datePart} ${timePart} IST`;
}

/**
 * Calculate duration between two timestamps
 * @param startMs - Start timestamp in milliseconds
 * @param endMs - End timestamp in milliseconds
 * @returns Formatted duration: "2h 15m 30s" or "10m 30s" or "45s"
 */
export function formatDuration(startMs: number, endMs: number): string {
  const diffMs = endMs - startMs;
  const totalSeconds = Math.floor(diffMs / 1000);
  
  const hours = Math.floor(totalSeconds / 3600);
  const minutes = Math.floor((totalSeconds % 3600) / 60);
  const seconds = totalSeconds % 60;
  
  if (hours > 0) {
    return `${hours}h ${minutes}m ${seconds}s`;
  } else if (minutes > 0) {
    return `${minutes}m ${seconds}s`;
  } else {
    return `${seconds}s`;
  }
}

/**
 * Get current IST timestamp
 * @returns Current timestamp in milliseconds
 */
export function nowIST(): number {
  return Date.now();
}

/**
 * Format current time to IST string
 * @returns Current time formatted as IST string
 */
export function currentIST(): string {
  return formatIST(nowIST());
}
